scrapy/utils/robotparser.py

```diff
@@ -0,0 +1,254 @@
+# this code is adapted from Python 3.x master branch
+""" robotparser.py
+
+    Copyright (C) 2000  Bastian Kleineidam
+
+    You can choose between two licenses when using this package:
+    1) GNU GPLv2
+    2) PSF license for Python 2.2
+
+    The robots.txt Exclusion Protocol is implemented as specified in
+    http://www.robotstxt.org/norobots-rfc.txt
+
+    Changes by Scrapy developers:
+    - Adapted for dual-Python2/Python3 support
+"""
+
+import collections
+try:
+    from urllib.parse import urlparse, urlunparse, quote, unquote
+except ImportError:
+    from urlparse import urlparse, urlunparse
+    from urllib import quote, unquote
+
+from w3lib.url import safe_url_string, canonicalize_url
+
+
+__all__ = ["RobotFileParser"]
+
+class RobotFileParser:
+    """ This class provides a set of methods to read, parse and answer
+    questions about a single robots.txt file.
+
+    """
+
+    def __init__(self, url=''):
+        self.entries = []
+        self.default_entry = None
+        self.disallow_all = False
+        self.allow_all = False
+        self.set_url(url)
+        self.last_checked = 0
+
+    def mtime(self):
+        """Returns the time the robots.txt file was last fetched.
+
+        This is useful for long-running web spiders that need to
+        check for new robots.txt files periodically.
+
+        """
+        return self.last_checked
+
+    def modified(self):
+        """Sets the time the robots.txt file was last fetched to the
+        current time.
+
+        """
+        import time
+        self.last_checked = time.time()
+
+    def set_url(self, url):
+        """Sets the URL referring to a robots.txt file."""
+        self.url = url
+        self.host, self.path = urlparse(url)[1:3]
+
+    def read(self):
+        """Reads the robots.txt URL and feeds it to the parser."""
+        raw = f.read()
```

Comment:
oh right. sorry about that. Yes, it's a leftover I haven't removed completely (it's the part that downloads robots.txt from URLs)